server/src/main/java/org/elasticsearch/index/codec/DeduplicatingFieldInfosFormat.java

```diff
@@ -0,0 +1,94 @@
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.index.codec;
+
+import org.apache.lucene.codecs.FieldInfosFormat;
+import org.apache.lucene.index.FieldInfo;
+import org.apache.lucene.index.FieldInfos;
+import org.apache.lucene.index.SegmentInfo;
+import org.apache.lucene.store.Directory;
+import org.apache.lucene.store.IOContext;
+import org.elasticsearch.common.util.Maps;
+import org.elasticsearch.common.util.StringLiteralDeduplicator;
+import org.elasticsearch.common.util.concurrent.ConcurrentCollections;
+import org.elasticsearch.index.mapper.FieldMapper;
+
+import java.io.IOException;
+import java.util.Map;
+
+/**
+ * Wrapper around a {@link FieldInfosFormat} that will deduplicate and intern all field names, attribute-keys and -values, and in most
+ * cases attribute maps on read. We use this to reduce the per-field overhead for Elasticsearch instances holding a large number of
+ * segments.
+ */
+public final class DeduplicatingFieldInfosFormat extends FieldInfosFormat {
+
+    private static final Map<Map<String, String>, Map<String, String>> attributeDeduplicator = ConcurrentCollections.newConcurrentMap();
+
+    private static final StringLiteralDeduplicator attributesDeduplicator = new StringLiteralDeduplicator();
+
+    private final FieldInfosFormat delegate;
+
+    public DeduplicatingFieldInfosFormat(FieldInfosFormat delegate) {
+        this.delegate = delegate;
+    }
+
+    @Override
+    public FieldInfos read(Directory directory, SegmentInfo segmentInfo, String segmentSuffix, IOContext iocontext) throws IOException {
+        final FieldInfos fieldInfos = delegate.read(directory, segmentInfo, segmentSuffix, iocontext);
+        final FieldInfo[] deduplicated = new FieldInfo[fieldInfos.size()];
+        int i = 0;
+        for (FieldInfo fi : fieldInfos) {
+            deduplicated[i++] = new FieldInfo(
+                FieldMapper.internFieldName(fi.getName()),
+                fi.number,
+                fi.hasVectors(),
+                fi.omitsNorms(),
+                fi.hasPayloads(),
+                fi.getIndexOptions(),
+                fi.getDocValuesType(),
+                fi.getDocValuesGen(),
+                // original attributes is UnmodifiableMap
+                internStringStringMap(fi.attributes()),
+                fi.getPointDimensionCount(),
+                fi.getPointIndexDimensionCount(),
+                fi.getPointNumBytes(),
+                fi.getVectorDimension(),
+                fi.getVectorEncoding(),
+                fi.getVectorSimilarityFunction(),
+                fi.isSoftDeletesField(),
+                fi.isParentField()
+            );
+        }
+        return new FieldInfos(deduplicated);
+    }
+
+    private static Map<String, String> internStringStringMap(Map<String, String> m) {
+        var res = attributeDeduplicator.get(m);
+        if (res == null) {
+            if (attributeDeduplicator.size() > 100) {
```

Comment:
Would the size of `attributeDeduplicator` map grow easily beyond 100? This is a static map.  I think within a shard there will be segments with different attributes. Maybe not that many, but this maybe a factor of the number shards of different data streams that get assigned to a node?

Reply:
I couldn't reproduce or find any case in the heap dumps I looked at over the last year or so. My thinking was that 100 is overkill already but doesn't hurt in any real sense and in case I'm wrong it prevents a catastrophic leak :)
The highest count of distinct attribute maps I could reproduce was 8 I think :)

Reply:
I think we are using only four attributes:

- PerFieldPostingsFormat.format
- PerFieldPostingsFormat.suffix
- PerFieldDocValuesFormat.format
- PerFieldDocValuesFormat.suffix

and 100 should be large enough here.

Reply:
ah thanks for confirming the 4 attributes situation! :)  that was my impression as well from the Lucene codebase :)

Reply:
Np, I think this map is likely to be bounded by 4 for each codec version.